channel/manager.go

```diff
@@ -331,38 +331,79 @@ func (inst *Instance) MscBaseState() MSCBaseStateSigned {
+	}
+
+	//Validate integrity of the peer signature on the state
+	isValidPeer, err := newState.VerifySign(inst.PeerID(), peerRole)
+
+	if err != nil {
+		return false, err.Error()
+	}
+	if !isValidPeer {
+		return false, "Invalid peer signature"
+	}
+
+	//when previous state exists, check if the current version number is greater than previous
+	lastVpcStateIndex := len(inst.vpcStatesList) - 1
+	if lastVpcStateIndex != -1 {
+		lastValidStateVersion := inst.vpcStatesList[lastVpcStateIndex].VPCState.Version
+		if newState.VPCState.Version.Cmp(lastValidStateVersion) != 1 {
+			return false, fmt.Sprintf("Current Version number (%s) less than previous (%s)", newState.VPCState.Version.String(), lastValidStateVersion.String())
+		}
+	}
+
+	return true, ""
+}
+
+// ValidateFullState validates the integrity of newState and if unsuccessful, returns the reason.
+// Version number, self and peer signatures are validated.
+func (inst *Instance) ValidateFullState(newState VPCStateSigned) (isValid bool, reason string) {
 
 	//Validate integrity of the sender signature on the state
 	isValidSender, err := newState.VerifySign(inst.SenderID(), Sender)
 	if err != nil {
-		return err
+		return false, "Invalid sender signature - " + err.Error()
 	}
 	if !isValidSender {
-		return fmt.Errorf("Sender signature on VPCState invalid")
+		return false, "Invalid sender signature"
 	}
 
 	//Validate integrity of the receiver signature on the state
 	isValidReceiver, err := newState.VerifySign(inst.ReceiverID(), Receiver)
 	if err != nil {
-		return err
+		return false, "Invalid receiver signature - " + err.Error()
 	}
 	if !isValidReceiver {
-		return fmt.Errorf("Receiver signature on VPCState invalid")
+		return false, "Invalid receiver signature"
 	}
 
-	if lastVpcStateIndex := len(inst.vpcStatesList) - 1; lastVpcStateIndex != -1 {
-
-		//when previous state exists, check if the current version number is greater than previous
+	//when previous state exists, check if the current version number is greater than previous
+	lastVpcStateIndex := len(inst.vpcStatesList) - 1
+	if lastVpcStateIndex != -1 {
 		lastValidStateVersion := inst.vpcStatesList[lastVpcStateIndex].VPCState.Version
 		if newState.VPCState.Version.Cmp(lastValidStateVersion) != 1 {
-			return fmt.Errorf("Current Version number (%s) less than previous (%s)", newState.VPCState.Version.String(), lastValidStateVersion.String())
+			return false, fmt.Sprintf("Current Version number (%s) less than previous (%s)", newState.VPCState.Version.String(), lastValidStateVersion.String())
 		}
 	}
-	logger.Debug("New MSC base state set")
+
+	return true, ""
+}
+
+// SetCurrentVPCState adds newState to vpc state list of the channel.
+// Validity of the state will not be checked, should be done prior to calling this function.
+func (inst *Instance) SetCurrentVPCState(newState VPCStateSigned) {
 	inst.vpcStatesList = append(inst.vpcStatesList, newState)
-	return nil
+	logger.Debug("New MSC base state set")
 }
```

Comment:
Currently I'm not sure if it makes sense to be able to add states without any validation.
Does it make sense to call here `ValidateFullState`?

Reply:
As stated in documentation of this function, the state should be validated before updating to the Instance. 
Because, two types of validation needs to be done : 
1. Integrity of the state and signatures over it.
2. Correctness of the state according to the specific application.
(Second part of validation will vary depending upon the specific application)

First part of the validation can be done by calling ValidFullState and second part needs to be implemented by the application and called before calling SetCurrentVPCState  .

Reply:
Point number 2 can only be done by the application or user, that's right.
Point number 1 is at least from my point of view mandatory, otherwise it does not make sense to add a state as valid. The other option is to distinguish between valid and intermediate, which is not treated as valid. Am I right that the `vpcStatesList` only holds valid states?

Reply:
Yes, you are right that `vpcStateList` will hold only valid states. 

If the correctness of the state according to application logic will be validated before adding the signature of the user (part 2), then a state with two signatures can be implicitly considered correct with respect to the application logic.

And it will be sufficient to check only the integrity of signatures over it before adding it to `vpcStateList`.

I will make changes accordingly and update the pull request.